### PR TITLE
separate tests that expects support on overlap config name and path

### PR DESCRIPTION
**Pull request by jieru-hu**
<!-- Thank you for sending a PR and taking the time to improve Hydra -->

## Motivation
The config source test suite expects the config source plugins to support the following patterns: config name and config path overlap.
```
/dataset.yaml
/dataset/imagenet.yaml
```
However this pattern is not supported in `configerator`, this PR aims to provide a way for the subclasses of this test suite to skip the tests when needed.

### Have you read the [Contributing Guidelines on pull requests](https://github.com/facebookresearch/hydra/blob/master/CONTRIBUTING.md)?

Yes

## Test Plan

created my own test_suite
**SKIP the tests**
```
@pytest.mark.parametrize(
    "type_, path",
    [
        pytest.param(
            FileConfigSource,
            "file://tests/test_apps/config_source_test/dir",
            id="FileConfigSource",
        ),
    ],
)
class TestCoreConfigSources(ConfigSourceTestSuite):

    def skip_overlap_config_path_name(self) -> bool:
        return True
```
output

```
$ pytest -s /Users/jieru/workspace/hydra-fork/hydra/tests/test_jieru.py
====================================================================================================== test session starts =======================================================================================================
platform darwin -- Python 3.8.2, pytest-5.4.2, py-1.8.1, pluggy-0.13.1
rootdir: /Users/jieru/workspace/hydra-fork/hydra, inifile: pytest.ini
plugins: snail-0.1.0, hydra-core-1.1.0.dev1
collected 51 items                                                                                                                                                                                                               

tests/test_jieru.py ...................s................s..............

================================================================================================= 49 passed, 2 skipped in 0.11s ==================================================================================================
(pytest38) jieru-mbp:hydra jieru$ 

```

**Do not skip**


```
@pytest.mark.parametrize(
    "type_, path",
    [
        pytest.param(
            FileConfigSource,
            "file://tests/test_apps/config_source_test/dir",
            id="FileConfigSource",
        ),
    ],
)
class TestCoreConfigSources(ConfigSourceTestSuite):
    pass

```
output
```
$ pytest -s /Users/jieru/workspace/hydra-fork/hydra/tests/test_jieru.py
================================================================================================================================ test session starts ================================================================================================================================
platform darwin -- Python 3.8.2, pytest-5.4.2, py-1.8.1, pluggy-0.13.1
rootdir: /Users/jieru/workspace/hydra-fork/hydra, inifile: pytest.ini
plugins: snail-0.1.0, hydra-core-1.1.0.dev1
collected 51 items                                                                                                                                                                                                                                                                  

tests/test_jieru.py ...................................................

================================================================================================================================ 51 passed in 0.14s =================================================================================================================================
(pytest38) jieru-mbp:hydra jieru$ 

```
## Related Issues and PRs

(Is this PR part of a group of changes? Link the other relevant PRs and Issues here. Use https://help.github.com/en/articles/closing-issues-using-keywords for help on GitHub syntax)

### CODE REVIEW

File: hydra/test_utils/config_source_common_tests.py [hunk 1]
```diff
@@ -10,6 +11,15 @@
 
 
 class ConfigSourceTestSuite:
+    def skip_overlap_config_path_name(self) -> bool:
+        """
+        Some config source plugins do not support config name and path overlap, for example:
+        /dateset.yaml
+        /dataset/cifar.yaml
+        is not allowed in configerator.
```
Comment (omry) [inline hydra/test_utils/config_source_common_tests.py:19]: do not mention configerator in the code.
just say that it's not supported by some config source plugins.

File: hydra/test_utils/config_source_common_tests.py [hunk 2]
```diff
@@ -55,10 +72,35 @@ def test_is_group(
...
+        results_filter: Optional[ObjectType],
+        expected: List[str],
     ) -> None:
         src = type_(provider="foo", path=path)
-        ret = src.is_config(config_path=config_path)
-        assert ret == expected
+        ret = src.list(config_path=config_path, results_filter=results_filter)
+        for x in expected:
+            assert x in ret
+        assert ret == sorted(ret)
```
Comment (omry) [inline hydra/test_utils/config_source_common_tests.py:103]: those helpers are adding more code than they are helping you reduce.
maybe just copy repeat the few important lines in it twice?

Comment (jieru-hu) [reply hydra/test_utils/config_source_common_tests.py:103]: 😅 ok sounds good.

File: hydra/test_utils/config_source_common_tests.py (hunk 1 above)
Comment (jieru-hu) [reply hydra/test_utils/config_source_common_tests.py:19]: removed. :)

File: hydra/test_utils/config_source_common_tests.py [hunk 3]
```diff
@@ -60,6 +69,21 @@ def test_is_config(
...
+        "config_path, expected",
+        [
+            ("dataset", True),
+        ],
+    )
+    def test_is_config_with_overlap_name(
+        self, type_: Type[ConfigSource], path: str, config_path: str, expected: bool
+    ) -> None:
+        if self.skip_overlap_config_path_name():
+            pytest.skip("Plugin does not support overlap config name with config path.")
```
Comment (omry) [inline hydra/test_utils/config_source_common_tests.py:82]: add the plugin name (or type) to the error.
maybe something like:
```python
f"ConfigSourcePlugin {type_.__name__} does not support overlap config name with config path"
```

Also, this reads  better to me:
ConfigSourcePlugin X does not support config objects and config groups with overlapping names.

File: hydra/test_utils/config_source_common_tests.py [hunk 4]
```diff
@@ -10,6 +11,15 @@
 
 
 class ConfigSourceTestSuite:
+    def skip_overlap_config_path_name(self) -> bool:
+        """
+        Some config source plugins do not support config name and path overlap, for example:
```
Comment (omry) [inline hydra/test_utils/config_source_common_tests.py:16]: Extended the commend to explain what this function is and how it should be used.

File: hydra/test_utils/config_source_common_tests.py [hunk 5]
```diff
@@ -70,7 +94,6 @@ def test_is_config(
             ("level1", ObjectType.GROUP, ["level2"]),
             ("level1/level2", ObjectType.GROUP, []),
             # Configs
-            ("", ObjectType.CONFIG, ["config_without_group", "dataset"]),
```
Comment (omry) [inline hydra/test_utils/config_source_common_tests.py:73]: Unless there is an additional problem I am not aware of, `config_without_group` should not be skipped.

**Review by omry — CHANGES_REQUESTED** [review]

File: hydra/test_utils/config_source_common_tests.py [hunk 6]
```diff
@@ -13,11 +13,13 @@
...
-        Some config source plugins do not support config name and path overlap, for example:
+        Some config source plugins do not support config name and path overlap.
+        For example the following may not be allowed:
+        (dataset exists both as a config object and a config group)
         /dateset.yaml
         /dataset/cifar.yaml
-        may not be allowed.
-        """
+
+        Override this function to return True so the test assumes /dataset/cifar.yaml exists instead of /dataset.yaml"""
```
Comment (omry) [inline hydra/test_utils/config_source_common_tests.py:22]: what?
why does it mean cifar.yaml exists INSTEAD of dataset.yaml? those two things seems unrelated.

Comment (jieru-hu) [reply hydra/test_utils/config_source_common_tests.py]: in configerator's case that  is what we want to test to think right? (we cannot have dataset.yaml and /dataset/cifar.yaml both exist)

Comment (omry) [reply hydra/test_utils/config_source_common_tests.py]: so we exclude dataset.yaml (cconf) from the test, right?

Comment (jieru-hu) [reply hydra/test_utils/config_source_common_tests.py]: Yea,  that's the plan :) 
Are you suggesting a wording change here?

Comment (omry) [reply hydra/test_utils/config_source_common_tests.py]: yes. the comment there has nothing to do with how we are solving the problem as far as I can tell.

**Comment by omry** [comment]
[this](https://docs.pytest.org/en/latest/skipping.html#skip-xfail-with-parametrize) might actually simplify this change.

Maybe try to set some global flag (or env variable) to alternative between with and without overlapping name tests?
```python
@pytest.mark.parametrize(
    ("n", "expected"),
    [
        pytest.param(
            10, 11, marks=pytest.mark.skipif(sys.version_info >= (3, 0), reason="py2k")
        ),
    ],
)
def test_increment(n, expected):
    assert n + 1 == expected
```

File: hydra/test_utils/config_source_common_tests.py (hunk 6 above)
Comment (jieru-hu) [reply hydra/test_utils/config_source_common_tests.py]: Got it, will update the doc.

**Review by omry — APPROVED** [review]

**Comment by jieru-hu** [comment]
> [this](https://docs.pytest.org/en/latest/skipping.html#skip-xfail-with-parametrize) might actually simplify this change.
> 
> Maybe try to set some global flag (or env variable) to alternative between with and without overlapping name tests?
> 
> ```python
> @pytest.mark.parametrize(
>     ("n", "expected"),
>     [
>         pytest.param(
>             10, 11, marks=pytest.mark.skipif(sys.version_info >= (3, 0), reason="py2k")
>         ),
>     ],
> )
> def test_increment(n, expected):
>     assert n + 1 == expected
> ```

I spent some time yesterday on trying to get this to work, but I could not figure out a way to pass a env variable to the `pytest.param` without calling the test like`SKIP=1 pytest test_bleh.py` :(
